### DISCUSSION TITLE
About Camera position

### DISCUSSION THREAD

**Discussion by linhongbin** [surgical-robotics-ai/surgical_robotics_challenge#15] [Q&A] [answered]
Hi,

I have a few question on camera position control:

- Could we control the camera position during Challenge # 2 and # 3?
- If we could, then which topics can we use control? "/ambf/env/cameras/cameraL/Command" or "/CRTK/ecm/servo_jp"? However I found ecm control did not works after I sent ros command.
- If we could not, the camera is set to default position when the simulator starts. Will the default camera position change (vary) during evaluation of the challenge?

**Comment by adnanmunawar** [comment, marked as answer]
Dear Lin,

Thank you for your questions. Here are the answers to them:

> Could we control the camera position during Challenge # 2 and # 3?

Yes, the camera can be (and expected to be) moved during all the 3 tasks.
 
> If we could, then which topics can we use control? "/ambf/env/cameras/cameraL/Command" or "/CRTK/ecm/servo_jp"? However I found ecm control did not works after I sent ros command.

To control the camera, you would use either the `/CRTK/ecm/servo_jp` topic or the `servo_jp` method from the [`ecm_arm.py`](https://github.com/collaborative-robotics/surgical_robotics_challenge/blob/master/scripts/surgical_robotics_challenge/ecm_arm.py) and **NOT** the `/ambf/env/cameras/cameraL/Command`. 

> However I found ecm control did not works after I sent ros command.

Thank you for reporting this, we shall look into this issue.

> If we could not, the camera is set to default position when the simulator starts. Will the default camera position change (vary) during evaluation of the challenge?

Yes, you can control the camera, and the default (starting) camera pose may change during the challenge.

**Reply by linhongbin** [reply]
Hi adnan,

Thanks for your quick response. Appreciate it.

I re-tested the ecm control and found the topic `/CRTK/ecm/servo_jp` was working. I was previously confused by using `/CRTK/ecm/servo_cp` instead. (I think the there is good reason for not having cartesian pose control since there are only 4 DOFs for ECM, and we cannot arbitrarily set ecm to any 6DOF pose).

I still have two follow-up questions: 
- Do we need to set joint limit when moving  ECM joints? Since I found I can arbitrarily set ecm joint position to any values.
>Yes, you can control the camera, and the default (starting) camera pose may change during the challenge.
- Does it means during challenge evaluation, the ecm joint positions (pitch, row, insertion, yaw)  are randomly set? Or even the base of ecm is randomly set? The first case can be easily solved by moving ecm joints to some fixed position before starting control.

**Reply by adnanmunawar** [reply]
Hi Lin,

These are the answers to your questions:

> Do we need to set joint limit when moving ECM joints? Since I found I can arbitrarily set ecm joint position to any values.

I think this might a good idea. Currently there are no joint limits that are enforced, but we might add these limits in the code soon.

> Does it means during challenge evaluation, the ecm joint positions (pitch, row, insertion, yaw) are randomly set? Or even the base of ecm is randomly set? The first case can be easily solved by moving ecm joints to some fixed position before starting control.

The ECM itself is just a frame, and it does not have a robot arm attached to it. Thus the joint values (pitch, row, insertion, yaw) are only for mimicking the control. The "base" of the ECM is another frame called "cameraFrame", which is always going to be fixed. The thing that would change is essentially the "ECM" frame which could be achieved by changing the pitch, row, insertion, and yaw values. However, you cannot retrieve the values of pitch, row, insertion, and yaw as these aren't actual joints, but you can directly get the pose of the "ECM" frame. That being said, we may add the reporting of joint value to the repo.

**Reply by linhongbin** [reply]
Hi Adnan, Sorry to bother you again. As you mentioned that "we might add these limits in the code soon." Do we expect enforced joint limits for ECM during the evaluation? Just want to doubly check the details.

**Reply by adnanmunawar** [reply]
Hi Hongbin, 
Thanks for the question. We did not implement the joint limits for the ECM so you should not expect it for the purpose of the challenge :).

**Reply by linhongbin** [reply]
Great, thanks for your detailed reply. Have a nice day.

**Comment by linhongbin** [comment]
There is a minor on topics `/CRTK/ecm/meared_js`. It kept publishing [0,0,0,0] for joint position. But it is not a big deal and I can use other signals instead.

**Reply by adnanmunawar** [reply]
I tried answering the question above.
